🗒️ Add notes about implementation of Winternitz signing (#11)
distributed-lab/nero: docs/paper/bitvm2_fast_recap.tex

$\sigma_{j-1}$ to the stack in the \texttt{scriptSig}, making the
\texttt{scriptSig} of the transaction input like this:

\begin{empheqboxed}
  \begin{align*}
    &\elem{z_{j-1}} \opcode{OP\_DUP} \elem{\sigma_{j-1}}
    \elem{\mathsf{pk}_{j-1}} \opcode{OP\_WINTERNITZVERIFY} \\
    &\elem{z_{j}} \opcode{OP\_DUP} \elem{\sigma_{j}}
    \elem{\mathsf{pk}_{j}} \opcode{OP\_WINTERNITZVERIFY} \\
    &\elem{f_j} \opcode{OP\_EQUAL} \opcode{OP\_NOT}
  \end{align*}
\end{empheqboxed}

where \texttt{OP\_WINTERNITZVERIFY} is the verification of the
Winternitz signature (commitment), described in Bitcoin Script (as
  Bitcoin Script does not have a built-in \texttt{OP\_CODE} for
Winternitz signatures)\footnote{Its implementation can be found here:
\url{https://github.com/distributed-lab/bitvm2-splitter/blob/feature/winternitz/bitcoin-winternitz/src/lib.rs}.}.

The correct implementation of \texttt{OP\_WINTERNITZVERIFY} in Bitcoin
Script would require the $\mathsf{Enc}(m)$ to split the state into $d$
digit numbers or recover the state from $d$ digit numbers on stack,
which without the bitwise operations or \texttt{OP\_CAT} for larger
than 32-bit values is nearly impossible.

> \subsubsection{32-bit stack elements}\label{sec:winternitz-31-bit-integers}

Reviewer: I would rather call this section "Winternitz Signatures meet Bitcoin" or "Winternitz Signatures in Bitcoin Script" or something similar.

Instead of bitwise operations, script can use arithmetic ones, but the

> Bitcoin Script supports opcodes such as \texttt{OP\_ADD}, only for

Reviewer: ```suggestion
Instead of bitwise operations, the Bitcoin script can use arithmetic ones. Still, this arithmetic is limited and contains only basic opcodes such as `OP_ADD`. To make matters worse, all the corresponding operations can be applied to 32-bit elements only.
```

32-bit signed integers, and as the last one is reserved for a sign,
only 31 bits can be used to store the state. This could be considered
as limitation, but as most of the complex arithmetic operations in
BitVM are done using only 32-bit integers, we can easily fix $\ell$ as

> $\ell = 32$ for all intermidiate states.

Reviewer: Did not understand what you mean in this sentence

By defining the $\mathsf{Enc}(m)$ as a domination free function

> $P(m)$, like described in~\cite{applied-crypto}, $d+1$ becomes the

Reviewer: ```suggestion
$P(m)$, like described in~\cite{applied-crypto}, it is convenient to make $d+1$ the power of two. Therefore, from now on, we set $d+1 = 2^w$ for some $w \in \mathbb{N}$.    
```

power of two which splits $m$ by some number of equal chunks of
$\log_2{(d+1)}$ bits, and $k$ becomes the sum of $n_0$ --- the number of
$d$-digit numbers from $m$, and $n_1$ a checksum (see
table~\ref{tab:winternitz}).

> \iffalse{}

Author: This macro excludes python code below, but I found it useful for myself keeping it somewhere. Not sure should we move it somewhere or leave it as it is

Reviewer: We will add the pseudocode for that later (using `algorithm2e` package)

%The python script i used for this table:
\begin{verbatim}
import math
l = 32
d = [3, 7, 15, 31, 63, 127, 255]
bits = lambda d: math.log(d+1, 2)
bits = lambda d: math.ceil(math.log(d+1, 2))
n0 = lambda bits: math.ceil(l / bits)
n1 = lambda bits,n0: math.ceil((2**(bits) * n0).bit_length() / bits)
k = lambda n0, n1: n0 + n1

for d in ds:
   bits_ = bits(d)
   n0_ = n0(bits_)
   n1_ = n1(bits_, n0_)
   k_ = k(n0_, n1_)
   print(f"{bits_} & {d} & {n0_} & {n1_} & {k_} \\\\")
\end{verbatim}
\fi

\begin{table}[h]
  \centering
  \begin{tabular}{lcccc}
    \toprule

>     bits & $d$ & $n_0$ & $n_1$ & $k$ \\

Reviewer: Merge columns 'bits' and ' $d$ ' into a single one with the parameter $w$ (defined in my comments above)

    \midrule
    2 & 3 & 16 & 4 & 20 \\
    3 & 7 & 11 & 3 & 14 \\
    4 & 15 & 8 & 2 & 10 \\
    5 & 31 & 7 & 2 & 9 \\
    6 & 63 & 6 & 2 & 8 \\
    7 & 127 & 5 & 2 & 7 \\
    8 & 255 & 4 & 2 & 6 \\
    \bottomrule
  \end{tabular}
  \caption{Different values of $k$ depending on $d$ for 32-bit
  message}\label{tab:winternitz}
\end{table}

> If chunks are equal lengthes, the recovery from $n_0$ digits is

Reviewer: ```suggestion
If chunks are of equal lengths, the recovery from $n_0$ digits is
```

> simply as:

Reviewer: ```suggestion
simply:
```

\begin{equation}

>   m = \sum_{i = 0}^{n_0} e_i \cdot 2^{i \log_2{(d+1)}}

Reviewer: Just set $w := \log_2(d+1)$ to avoid this lengthy notation :)

  \label{eq:u32-recovery}
\end{equation}

Where multiplication by powers of two can be implemented in Bitcoin
Script with sequence of \texttt{OP\_DUP} and \texttt{OP\_ADD}
opcodes. So, for example, multiplication of $e_j$ by $2^n$ in Bitcoin
Script is:

\begin{empheqboxed}
  \begin{align*}

>     \elem{e_j} \underbrace{\opcode{OP\_DUP} \opcode{OP\_ADD}}_{n \; times}

Reviewer: ```suggestion
    \elem{e_j} \underbrace{\opcode{OP\_DUP} \opcode{OP\_ADD}}_{n \; \text{times}}
```

  \end{align*}
\end{empheqboxed}

As Bitcoin Script has no loops or jumps, implementing dynamic number
of operations, like hashing something $d - e_j$ times without knowing
the $e_j$ before hand is challenging. That's why implementation uses
the ``lookup'' table of all $d$ hashes of signature's part $\sigma_j$ and
by using \texttt{OP\_PICK} pop the $d - e_j$ one on the top of stack,
like this:

\begin{empheqboxed}
  \begin{align*}
    & \elem{\sigma_j} \underbrace{\opcode{OP\_DUP}

>     \opcode{OP\_HASH}}_{d \; times} \\

Reviewer: ```suggestion
    \opcode{OP\_HASH}}_{d \; \text{times}} \\
```

    & \elem{e_j} \opcode{OP\_PICK}
  \end{align*}
\end{empheqboxed}

Still the upper bound for script size in Bitcoin persists, but the
current implementation requires around 1000 bytes per 32-bit stack
element, which is unfortunatly a lot. Parts of the public key make the
largest contribution to the script size. Assuming that as $H$
implementation uses \texttt{OP\_HASH160}, each part $(y_1,\dots,y_k)$
of the public key $\mathsf{pk}_{m}$ adds 40 bytes to the total script
size. Additionaly, for calculating the lookup table for signature
verification, $2d \cdot k$ opcodes are used. Further more, for message
recovery $2 \sum_{i = 0}^{n_0} i \log_2{(d+1)}$ number of opcodes are
added too. So the total script size, excluding utility opcodes, will

> be at least $40 * k + 2 d k + 2 \sum_{i = 0}^{n_0} i \log_2{(d+1)}$ (the

Reviewer: ```suggestion
be at least $40k + 2 d k + 2 \sum_{i = 0}^{n_0} i \log_2{(d+1)}$ (the
```

Reviewer: Also note that 

$$
2\sum_{i=0}^{n_0} i \log_2(d+1) =  \log_2(d+1)n_0(n_0+1) \approx wn_0^2
$$
